### PR TITLE
RI-8326 Collapse large vector embeddings in the query editor

### PR DESCRIPTION
**Pull request by dantovska**
# What

Collapse each large vector embedding detected in the query editor (Workbench and Vector Search) into a compact, clickable chip — `▸ vector · N dims` plus a copy button — so long FLOAT32 blobs no longer dominate the editor.

- Detected embeddings auto-collapse (a pasted query starts collapsed); click the chip to expand, the `▾` arrow to re-collapse.
- No data loss: the blob is replaced with a short placeholder in the model while the full value is kept in an in-memory store and restored at every exit path — run/submit, copy, and save-to-library.
- Copy button copies the full embedding without expanding; hover shows the embedding metadata (dimensions + byte size) on the label and a `Copy` tooltip on the button.
- Chips are styled after the selected "Query editor" tab and adapt to light/dark; clicking a chip never leaves the caret on it.

Builds on the detection layer from RI-8317 (#6215, merged).

# Testing

1. Open Vector Search (or Workbench) and paste/run a query containing a large vector embedding — a binary `\x` FLOAT32 blob or a numeric array with ≥ 10 elements (e.g. an `FT.SEARCH … PARAMS 2 my_blob "<blob>"`).
2. The embedding collapses to a chip, the rest of the query stays in place, and the caret returns to the top.
3. Click the chip to expand and the `▾` arrow to collapse again; click the copy button to copy the full value; hover the label for metadata and the copy button for the `Copy` tooltip.
4. Run the query and copy the whole query — the full embedding value is used (no placeholder leaks into the command or clipboard).
5. Verify in both light and dark themes.

Unit tests cover the detection key helper, the placeholder store, and the query-editor context expansion (`npm test`).

---

Closes #RI-8326

---

> [!NOTE]
> **Medium Risk**
> Query run and save paths depend on correct placeholder expansion; a bug could send truncated or wrong embedding text to Redis, though behavior is covered by tests and failures leave unknown placeholders visible.
> 
> **Overview**
> **Collapses detected large vector embeddings** in Workbench and Vector Search query editors into compact chips (`▸ vector · N dims` plus copy), wired through new `useVectorEmbeddingCollapse` on `VectorEmbeddingHighlight`.
> 
> The editor model holds short **session-scoped placeholders** (`vectorEmbeddingPlaceholders`); full blob text lives in an in-memory map keyed by unguessable ids. Users can expand/collapse via chip UI, copy without expanding, and undo/redo is handled so re-collapsing does not fight Ctrl+Z. **Copy/cut** rewrites clipboard text to expanded values; **submit** (`QueryEditorContext`) and **query-library save** (`QueryEditorWrapper`) call `expandVectorEmbeddings` so placeholders never reach Redis or persisted queries.
> 
> Adds `getEmbeddingKey` so PARAMS-named embeddings stay expanded while editing, chip **global styles**, i18n labels, Monaco test mocks, and unit tests for the hook, placeholder store, and detection edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 68567751233d1e4c9fd5f962aa9f250a41ef43d1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+      const expanded = expandVectorEmbeddings(selected)
+      if (expanded === selected) return
+
+      e.preventDefault()
+      e.stopPropagation()
+      e.clipboardData.setData('text/plain', expanded)
+    }
+
+    domNode.addEventListener('mousedown', handleChipMouseDown, true)
+    domNode.addEventListener('copy', handleCopyEvent, true)
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:335]: Handled: the hook registers a capture-phase `cut` listener (`handleCutEvent` → `writeExpandedClipboard(e, true)`) that expands placeholders to their full values before the clipboard write and still performs the deletion. Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  id: number,
+  dimensions: number,
+): string => `[▸ vector · ${dimensions} dims #${id}]`
+
+export const collapseVectorEmbeddingValue = (
+  value: string,
+  dimensions: number,
+  byteSize: number,
+): string => {
+  const id = nextPlaceholderId
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts]: Fixed: placeholder ids are now session-scoped strings (`<session>-<n>`) and the regex requires that shape, so a bare or foreign token like `#1` can never match a live id (e56373f5f, hardened in 78c03b52a with a crypto-seeded session). Added unit tests for a bare token and a foreign placeholder. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+      const selection = editor.getSelection()
+      editor.executeEdits(COLLAPSE_EDIT_SOURCE, [
+        {
+          range: toMonacoRange(monaco, currentModel, placeholder.range),
+          text: value,
+        },
+      ])
+      // The value is back in the editor; the placeholder id is now dead.
+      releaseVectorEmbeddingValue(placeholder.id)
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts]: Fixed in 78c03b52a: the stored value is no longer released when a chip is expanded, so an undo that restores the placeholder text still resolves on expand/copy/submit. The store is session-scoped and cleared on reset. Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+const PLACEHOLDER_REGEX = /\[▸ vector · (\d+) dims #([a-z0-9]+-\d+)\]/g
+
+interface StoredEmbedding {
+  value: string
+  byteSize: number
+}
+
+const newSessionId = (): string => Math.random().toString(36).slice(2, 8)
+
+let sessionId = newSessionId()
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts:12]: Fixed in 78c03b52a: the session id now derives from `crypto.getRandomValues` instead of `Math.random()`. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    // Drop expanded keys whose embedding is no longer in the editor (cleared,
+    // replaced, deleted) so a later paste always starts collapsed.
+    const detectedKeys = new Set(detected.map(getEmbeddingKey))
+    userExpandedKeys.current = new Set(
+      [...userExpandedKeys.current].filter((key) => detectedKeys.has(key)),
+    )
+
+    const marksToCollapse = detected.filter(
+      (mark) => !userExpandedKeys.current.has(getEmbeddingKey(mark)),
+    )
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:96]: Fixed: `getEmbeddingKey` now prefers the PARAMS name over a content signature. The name is stable across edits to the blob value, so an expanded embedding keeps its identity and is no longer re-collapsed while the user types. Added a unit test. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:96]: Fixed by the same change: keying by PARAMS name makes distinct entries distinct, so pasting a duplicate embedding under a different param name still collapses even while another copy is expanded. Falls back to a content signature only for unnamed embeddings (raw array / non-FT query). Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  return buildVectorEmbeddingPlaceholder(id, dimensions)
+}
+
+export const getVectorEmbeddingValue = (id: string): string | undefined =>
+  collapsedValues.get(id)?.value
+
+// Frees the stored value once its placeholder is gone (e.g. after expanding).
+export const releaseVectorEmbeddingValue = (id: string): void => {
+  collapsedValues.delete(id)
+}
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts]: Fixed: the store is now capped (`MAX_STORED_EMBEDDINGS`) with least-recently-used eviction. Placeholders still present in a model are read on every render, so they stay hot; only stale (removed/cleared) values age out, without risking expand/copy/submit for active ones. Added unit tests. Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+): string => `[▸ vector · ${dimensions} dims #${id}]`
+
+export const collapseVectorEmbeddingValue = (
+  value: string,
+  dimensions: number,
+  byteSize: number,
+): string => {
+  const id = nextPlaceholderId
+  nextPlaceholderId += 1
+  collapsedValues.set(id, { value, byteSize })
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts:28]: Addressed together with the LRU cap on the store (`MAX_STORED_EMBEDDINGS` + least-recently-used eviction). Values are no longer retained unbounded across a long session. Note: eager release on expand was intentionally dropped so an undo that restores a placeholder still resolves. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+  // Placeholder id whose copy button currently shows the "copied" tick.
+  const [copiedId, setCopiedId] = useState<Nullable<number>>(null)
+  const decorationCollection =
+    useRef<Nullable<monacoEditor.editor.IEditorDecorationsCollection>>(null)
+  const removeDomListeners = useRef<Nullable<() => void>>(null)
+  const userExpandedKeys = useRef<Set<string>>(new Set())
+  const copiedTimer = useRef<Nullable<ReturnType<typeof setTimeout>>>(null)
+
+  useEffect(() => {
+    if (!monacoObjects.current) return
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:74]: Fixed: the hook now flips an `isEditorReady` flag once `monacoObjects.current` is attached (polled via requestAnimationFrame) and depends on it, so the initial collapse pass runs even when editorDidMount lands after the first render and the query never changes. Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingUtils.ts [hunk 10]
```diff
@@ -67,6 +67,14 @@ const buildParamNameLookup = (query: string): Map<string, string> => {
   return lookup
 }
 
+/**
+ * Content identity for an embedding, stable across edits elsewhere in the
+ * query. Keyed by value (not the PARAMS name) so two different embeddings that
+ * share a name are still distinct.
+ */
+export const getEmbeddingKey = (mark: VectorEmbeddingMark): string =>
+  `${mark.format}:${mark.dimensions}:${mark.firstValues.join(',')}:${mark.lastValues.join(',')}`
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingUtils.ts:79]: Scoped trade-off: this fallback content signature is only reached for embeddings that are **not** named PARAMS values (a raw numeric array, or a non-FT query). Named PARAMS embeddings — the vector-search / Workbench flow this feature targets — now key by the stable param name, so they are no longer re-collapsed while edited nor share expansion state. Fully covering the unnamed-fallback case needs range/placeholder-identity (Monaco sticky-decoration) tracking, whose edit-survival behaviour cannot be verified outside the running app; that redesign was deferred in favour of the param-name key. Resolving as an accepted, documented limitation.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  value: string,
+  dimensions: number,
+  byteSize: number,
+): string => {
+  const id = `${sessionId}-${nextPlaceholderId}`
+  nextPlaceholderId += 1
+  collapsedValues.set(id, { value, byteSize })
+  if (collapsedValues.size > MAX_STORED_EMBEDDINGS) {
+    const oldest = collapsedValues.keys().next().value
+    if (oldest !== undefined) collapsedValues.delete(oldest)
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts]: Good catch — reverted the cap in a3babd7c4. You are right that a blind count cap can evict values whose placeholders are still live (a query with more than the cap of embeddings, or the other query editor that shares this module-level store), which would send the literal placeholder to Redis on submit/save. Correctness wins: every stored value is kept for the session. A safe bound needs per-editor store scoping (so eviction can be limited to values with no live placeholder in that editor), which I have deferred to a follow-up. Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts]: Follow-up: the LRU cap introduced here was reverted in a3babd7c4 — it risked evicting values whose placeholders are still live (see the P2 thread on vectorEmbeddingPlaceholders.ts). The store is now kept for the session; a safe memory bound needs per-editor store scoping and is deferred to a follow-up.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts:28]: Follow-up: same as above — the store cap was reverted in a3babd7c4 to avoid evicting live placeholder values. Memory is session-bounded; a correct bound requires per-editor store scoping, deferred to a follow-up.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+): string => `[▸ vector · ${dimensions} dims #${id}]`
+
+export const collapseVectorEmbeddingValue = (
+  value: string,
+  dimensions: number,
+  byteSize: number,
+): string => {
+  const id = `${sessionId}-${nextPlaceholderId}`
+  nextPlaceholderId += 1
+  collapsedValues.set(id, { value, byteSize })
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts:28]: Acknowledged and tracked as a deferred limitation. The value is intentionally retained for the session: a bounded/GC approach that *preserves active placeholders* must know which ids still have a live placeholder across every editor sharing this module-level store (and keep a small undo cache so an undo that restores a placeholder still resolves). That is a scoped refactor of the store (per-editor scoping / live-placeholder registry) whose multi-editor + undo behaviour cannot be verified outside the running app, so it is out of scope for this PR and will be handled in a follow-up. In practice the store is session-bounded (cleared on reset), and the earlier naive cap was reverted because it evicted live values and corrupted submitted queries. Resolving as deferred.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    new RegExp(PLACEHOLDER_REGEX),
+    (placeholder, _dimensions: string, id: string) =>
+      collapsedValues.get(id)?.value ?? placeholder,
+  )
+
+export const resetVectorEmbeddingPlaceholders = (): void => {
+  sessionId = 'sess'
+  nextPlaceholderId = 1
+  collapsedValues.clear()
+}
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts:70]: Fixed in 6e196eaf1: `resetVectorEmbeddingPlaceholders` now regenerates a CSPRNG-based session id (same as module init) instead of the hardcoded `'sess'`, so the unguessable-prefix property holds for any caller. Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  dimensions: number
+  /** Undefined when the value is unknown (e.g. a query from another session). */
+  byteSize?: number
+  range: VectorEmbeddingRange
+}
+
+export const buildVectorEmbeddingPlaceholder = (
+  id: string,
+  dimensions: number,
+): string => `[▸ vector · ${dimensions} dims #${id}]`
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingPlaceholders.ts]: Fixed in 5ff5cb4db: the model placeholder is now space-free (`[▸vector·Ndims#id]`), so the FT.SEARCH tokenizer treats a collapsed value as a single PARAMS argument and later embeddings keep their param-name mapping (no fallback-to-content re-collapse). The visible chip label with spaces is drawn as a separate injected decoration and is unchanged. Added a regression test asserting a later PARAMS name survives an earlier value being collapsed. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+    // Drop expanded keys whose embedding is no longer in the editor (cleared,
+    // replaced, deleted) so a later paste always starts collapsed.
+    const detectedKeys = new Set(detected.map(getEmbeddingKey))
+    userExpandedKeys.current = new Set(
+      [...userExpandedKeys.current].filter((key) => detectedKeys.has(key)),
+    )
+
+    const marksToCollapse = detected.filter(
+      (mark) => !userExpandedKeys.current.has(getEmbeddingKey(mark)),
+    )
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:96]: Fixed in the latest commit: the hook now watches the editor content change event and, when it is an undo/redo (`isUndoing`/`isRedoing`), skips the auto-collapse pass once so the restored raw embedding is left in place and Ctrl+Z can step back past a collapse. A later genuine edit still collapses as before. (Undo/redo event timing is worth a quick in-app smoke.) Resolving.

File: redisinsight/ui/src/utils/monaco/vectorEmbeddingUtils.ts [hunk 16]
```diff
@@ -67,6 +67,17 @@ const buildParamNameLookup = (query: string): Map<string, string> => {
...
+/**
+ * Stable identity for an embedding across query edits. Prefer the PARAMS name:
+ * it survives edits to the blob's own value (so an expanded embedding is not
+ * re-collapsed while the user types) and is distinct per PARAMS entry (so a
+ * pasted duplicate still collapses). Falls back to content when the embedding
+ * is not a named PARAMS value (raw array, non-FT query).
+ */
+export const getEmbeddingKey = (mark: VectorEmbeddingMark): string =>
+  mark.paramName ??
+  `${mark.format}:${mark.dimensions}:${mark.firstValues.join(',')}:${mark.lastValues.join(',')}`
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/utils/monaco/vectorEmbeddingUtils.ts:79]: Acknowledged as a known, deferred edge case. `buildParamNameLookup` is pre-existing RI-8317 code (merged in #6215) that maps only the first PARAMS clause; this PR depends on `paramName` for the collapse key but does not own that parser, so extending it to multiple commands is out of scope here. More broadly, this is one facet of a single limitation: expansion state is keyed by param-name with a content-signature fallback, so any gap in param-name coverage (unnamed values, multi-command scripts, FT.AGGREGATE, …) drops to the content key and can re-collapse on edit. The single-command PARAMS flow this feature targets works. The complete fix is to anchor expansion state to a sticky Monaco decoration/range rather than a name/content key — a follow-up, since it needs verification in the running app. Resolving as deferred.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+    // An undo/redo restored raw embedding text; don't re-collapse it this pass,
+    // or Ctrl+Z could never step back past a collapse. Consume the flag so the
+    // next genuine edit collapses as usual.
+    const isUndoRestore = skipAutoCollapseOnce.current
+    skipAutoCollapseOnce.current = false
+    if (marksToCollapse.length > 0 && !isUndoRestore) {
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts]: Good catch — fixed in the latest commit. Instead of only skipping one pass (which an unrelated re-render like the copied-tick timer could slip past), an undo/redo-restored embedding is now added to `userExpandedKeys`, so it is treated as user-expanded and stays put across later re-renders until it is removed. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+    // than just skipping this one pass, so an unrelated later re-render (e.g.
+    // the copied-tick timer) does not re-collapse it and defeat the undo.
+    const isUndoRestore = skipAutoCollapseOnce.current
+    skipAutoCollapseOnce.current = false
+    if (marksToCollapse.length > 0 && isUndoRestore) {
+      marksToCollapse.forEach((mark) =>
+        userExpandedKeys.current.add(getEmbeddingKey(mark)),
+      )
+    } else if (marksToCollapse.length > 0) {
+      editor.executeEdits(
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:113]: Fixed in the latest commit: the auto-collapse `executeEdits` is now wrapped in explicit `pushUndoStop()` calls, so the collapse is a separate undo entry from the paste/type that produced the vector. Ctrl+Z now lands on the raw value first (where the isUndoing guard keeps it expanded) rather than merging past it. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+            mark.byteSize,
+          ),
+        })),
+      )
+      editor.pushUndoStop()
+      // Pasting scrolls to the end of the blob; once collapsed, go back to top.
+      const top = { lineNumber: 1, column: 1 }
+      editor.setPosition(top)
+      editor.revealPosition(top)
+      return
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:130]: Fixed in the latest commit: the auto-collapse no longer forces the cursor to line 1. It now captures the selection before the edit and restores it afterward (Monaco maps it across the edit), matching the manual expand/collapse handlers, so a collapse triggered mid-edit keeps the caret where the user is. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+      )
+      if (expandedMark)
+        userExpandedKeys.current.add(getEmbeddingKey(expandedMark))
+      if (selection) editor.setSelection(selection)
+    }
+
+    // Copy/cut a selection with full values instead of placeholders; cut also
+    // removes the selection since we take over the clipboard write.
+    const writeExpandedClipboard = (e: ClipboardEvent, isCut: boolean) => {
+      const selection = editor.getSelection()
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts]: Fixed in the latest commit: the copy/cut handler now uses `getSelections()` and processes every selection ordered top-to-bottom — expanding placeholders in each range, joining with newlines, and (on cut) deleting all ranges — so multi-cursor copy/cut no longer drops ranges or leaves raw placeholders. When no range contains a placeholder it falls through to Monaco native handling. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+      const offset = mouseTarget?.position
+        ? currentModel.getOffsetAt(mouseTarget.position)
+        : null
+
+      if (isExpandArrow) {
+        const mark = findAtOffset(detectVectorEmbeddings(currentText), offset)
+        if (!mark) return
+        const selection = editor.getSelection()
+        userExpandedKeys.current.delete(getEmbeddingKey(mark))
+        editor.executeEdits(COLLAPSE_EDIT_SOURCE, [
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts:241]: Fixed in the latest commit: both manual chip paths (▾-collapse and expand) now bracket their `executeEdits` with `pushUndoStop()`, symmetric with the auto-collapse path, so a click after editing the vector is its own undo entry and Ctrl+Z no longer skips the intermediate state. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 22]
```diff
@@ -0,0 +1,342 @@
...
+          range: toMonacoRange(monaco, model, mark.range),
+          text: collapseVectorEmbeddingValue(
+            text.slice(mark.range.start, mark.range.end),
+            mark.dimensions,
+            mark.byteSize,
+          ),
+        })),
+      )
+      editor.pushUndoStop()
+      if (selection) editor.setSelection(selection)
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts]: Fixed in the latest commit: the auto-collapse no longer restores a captured selection (whose columns are stale once the long blob is replaced by a short placeholder). It lets `executeEdits` remap the caret next to the placeholder, then reveals that caret with `revealPositionInCenterIfOutsideViewport` — which also brings the view back after a paste scrolls to the blob end. Resolving.

File: redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+      ])
+      editor.pushUndoStop()
+      // Keep the stored value: an undo can restore the placeholder text, and it
+      // must still resolve on expand/copy/submit. The store is session-scoped.
+      const expandedMark = detectVectorEmbeddings(currentModel.getValue()).find(
+        (m) => m.range.start === placeholder.range.start,
+      )
+      if (expandedMark)
+        userExpandedKeys.current.add(getEmbeddingKey(expandedMark))
+      if (selection) editor.setSelection(selection)
```
[the remark this answers is not in the document]
Comment (dantovska) [reply redisinsight/ui/src/components/query/hooks/useVectorEmbeddingCollapse.ts]: Fixed in the latest commit for both manual paths (expand and ▾-collapse): dropped the pre-edit selection restore. The capture-phase handler already prevents the click from moving the caret, and executeEdits remaps the existing selection across the replacement, so the restore was redundant and its saved columns went stale once the text length changed. Resolving.

**Review by pawelangelow — APPROVED** [review]
